File: src/P0070_Climbing_Stairs.js
/**
 * @param {number} n
 * @return {number}
 */
var climbStairs = function(n) {
	if( n === 0) return 0;
	if(n === 1) return 1;
    var t = [1 , 1] , index = 2;
    while( index <= n){
    	t[index] = t[index - 1] + t[index - 2];
    	index++;
    }
    return t[n];
};

var n = 2;
var result = climbStairs(n);
console.log(result);

var n = 4;
var result = climbStairs(n);
console.log(result);

File: src/Java_Solutions/dp/P0070_Climbing_Stairs.java
package dp;

import java.util.HashMap;
import java.util.Map;

public class P0070_Climbing_Stairs {
    public static void main(String[] args) {
        P0070_Climbing_Stairs soln = new P0070_Climbing_Stairs();
        int n = 2;
        int result = soln.climbStairsRecursiveWithMemo(n);
        System.out.println(result);
        n = 4;
        result = soln.climbStairsRecursiveWithMemo(n);
        System.out.println(result);
    }

    public int climbStairs(int n) {
        if(n <=2){
            return n;
        }
        int[] dp = new int[n+1];
        dp[0] = 0;
        dp[1] = 1;
        dp[2] = 2;
        for(int i = 3 ; i <=n;i++){
            dp[i] = dp[i-1] + dp[i-2];
        }
        return dp[n];
    }

    public int climbStairsRecursiveWithMemo(int n) {
        return climbStairsHelper(n, new HashMap<>());

    }

    private int climbStairsHelper(int n , Map<Integer , Integer> mem) {
        int result;

        if (n == 1) {
            return 1;
        }

        if (n == 2) {
            return 2;
        }
        if(!mem.containsKey(n)) {
            result = climbStairsHelper(n - 1, mem) + climbStairsHelper(n - 2, mem);
            mem.put(n, result);
        }
        return mem.get(n);
    }
}
